backend/crud.py

```diff
@@ -0,0 +1,281 @@
+    try:
+        sch = Schema(name=data.name, slug=data.slug)
+        db.add(sch)
+        db.flush()
+    except sqlalchemy.exc.IntegrityError:
+        db.rollback()
+        raise SchemaExistsException(name=data.name, slug=data.slug)
+    
+    attr_names = set()
+    for attr in data.attributes:
+        if isinstance(attr, AttrDefSchema):
+            a: Attribute = db.execute(select(Attribute).where(Attribute.id == attr.attr_id)).scalar()
+            if a is None:
+                raise MissingAttributeException(attr.attr_id)
+        elif isinstance(attr, AttrDefWithAttrDataSchema):
+            a = create_attribute(db, attr, commit=False)
+            db.flush()
+        
+        if a.name in attr_names:
+            raise MultipleAttributeOccurencesException(a.name)
+        attr_names.add(a.name)
+
+        ad = AttributeDefinition(
+            attribute_id=a.id, 
+            schema_id=sch.id,
+            required=attr.required, 
+            list=attr.list, 
+            unique=attr.unique,
+            key=attr.key,
+            description=attr.description
+        )
+        db.add(ad)
+        db.flush()
+        if a.type == AttrType.FK:
+            if attr.bind_to_schema is None:
+                raise NoSchemaToBindException(attr_id=a.id)
+            if attr.bind_to_schema == -1:
+                s = sch
+            else:
+                s = db.execute(
+                    select(Schema)
+                    .where(Schema.id == attr.bind_to_schema)
+                    .where(Schema.deleted == False)
+                ).scalar()
+            if s is None:
+                raise MissingSchemaException(obj_id=attr.bind_to_schema)
+            bfk = BoundFK(attr_def_id=ad.id, schema_id=s.id)
+            db.add(bfk)
+    try:
+        db.commit()
+    except sqlalchemy.exc.IntegrityError:
+        db.rollback()
+        raise SchemaExistsException(name=data.name, slug=data.slug)
+    return sch
+
+
+def delete_schema(db: Session, schema_id: int) -> Schema:
+    schema = db.execute(
+        select(Schema)
+        .where(Schema.id == schema_id)
+        .where(Schema.deleted == False)
+    ).scalar()
+    if schema is None:
+        raise MissingSchemaException(obj_id=schema_id)
+    
+    db.execute(
+        update(Entity).where(Entity.schema_id == schema_id).values(deleted=True)
+    )
+    schema.deleted = True
+    db.commit()
+    return schema
+
+
+def update_schema(db: Session, schema_id: int, data: SchemaUpdateSchema) -> Schema:
+    sch: Schema = db.execute(
+        select(Schema).where(Schema.id == schema_id).where(Schema.deleted == False)
+    ).scalar()
+    if sch is None:
+        raise MissingSchemaException(obj_id=schema_id)
+    
+    schemas = db.execute(select(Schema).where( (Schema.name == data.name) | (Schema.slug == data.slug) )).scalars().all()
+    if len(schemas) == 1 and schemas[0].id != schema_id or len(schemas) > 1:
+        raise SchemaExistsException(name=data.name, slug=data.slug)
```

Comment:
We have defined the `name` and `slug` field as unique on the DB level. I'd suggest to rely SqlAlchemy to raise an exception and raise the custom exception from the `except:` block (as in `create_schema`).